nodist/clang_check_attributes.py

```diff
@@ -0,0 +1,325 @@
+#!/usr/bin/env python3
+# pylint: disable=C0114,C0115,C0116,C0209
+#
+# Copyright 2022 by Wilson Snyder. Verilator is free software; you
+# can redistribute it and/or modify it under the terms of either the GNU Lesser
+# General Public License Version 3 or the Apache License 2.0.
+# SPDX-License-Identifier: LGPL-3.0-only OR Apache-2.0
+#
+# Based on the work of (Guillaume "Vermeille" Sanchez)[https://github.com/Vermeille/clang-callgraph]
+# licensed under Apache 2.0
+
+import sys
+import os
+import glob
+import pathlib
+from pprint import pprint
+from clang.cindex import CursorKind, Index
+
+
+# https://stackoverflow.com/a/74205044
+class FixSizedDict(dict):
+
+    def __init__(self, *args, maxlen=0, **kwargs):
+        self._maxlen = maxlen
+        super().__init__(*args, **kwargs)
+
+    def __setitem__(self, key, value):
+        dict.__setitem__(self, key, value)
+        if self._maxlen > 0:
+            if len(self) > self._maxlen:
+                self.pop(next(iter(self)))
+
+
+INDEX = Index.create()
+CLANG_ARGS = []
+TRANSLATE_UNITS = FixSizedDict(maxlen=20)
+PRINTED = []
+
+# by definition constructor is MT safe,
+# but functions that constructor call are not!
+MT_SAFE_KINDS = [CursorKind.CONSTRUCTOR]
+UNSAFE_CALLS = []
```

Comment:
All caps for true constants is ok (e.g. VERILATOR_ROOT), but please use lower case with underscores for global variables as that is the Python way https://stackoverflow.com/questions/159720/what-is-the-naming-convention-in-python-for-variable-and-function